modules/internal/Bytes.chpl

```diff
@@ -0,0 +1,1634 @@
+   string.
+   */
+  class DecodeError: Error {
+    override proc message() {
+      return "Invalid UTF-8 character encountered.";
+    }
+  }
+
+  record bytes {  // _bytes when/if there is compiler support
+    pragma "no doc"
+    var len: int = 0; // length of string in bytes
+    pragma "no doc"
+    var _size: int = 0; // size of the buffer we own
+    pragma "no doc"
+    var buff: bufferType = nil;
+    pragma "no doc"
+    var isowned: bool = true;
+    pragma "no doc"
+    // We use chpl_nodeID as a shortcut to get at here.id without actually constructing
+    // a locale object. Used when determining if we should make a remote transfer.
+    var locale_id = chpl_nodeID; // : chpl_nodeID_t
+
+    pragma "no doc"
+    proc init() {
+
+    }
+
+    /*
+      Initialize a new :record:`bytes` from ``b``. If ``isowned`` is set to
+      ``true`` then ``b`` will be fully copied into the new instance. If it is
+      ``false`` a shallow copy will be made such that any in-place modifications
+      to the new bytes may appear in ``b``. It is the responsibility of the user
+      to ensure that the underlying buffer is not freed while being used as part
+      of a shallow copy.
+     */
+    /*proc init(b: bytes, isowned: bool = true) {*/
+
+    /*}*/
+
+    /*
+      Initialize a new :record:`bytes` from ``s``. If ``isowned`` is set to
+      ``true`` then ``b`` will be fully copied into the new instance. If it is
+      ``false`` a shallow copy will be made such that any in-place modifications
+      to the new bytes may appear in ``b``. It is the responsibility of the user
+      to ensure that the underlying buffer is not freed while being used as part
+      of a shallow copy.
+     */
+    proc init(s, isowned: bool = true) where s.type == string || s.type == bytes {
+      const sRemote = _local == false && s.locale_id != chpl_nodeID;
+      const sLen = s.len;
+      this.isowned = isowned;
+      this.complete();
+      // Don't need to do anything if s is an empty string
+      if sLen != 0 {
+        this.len = sLen;
+        if !_local && sRemote {
+          // ignore supplied value of isowned for remote strings so we don't leak
+          this.isowned = true;
+          this.buff = copyRemoteBuffer(s.locale_id, s.buff, sLen);
+          this._size = sLen+1;
+        } else {
+          if this.isowned {
+            const allocSize = chpl_here_good_alloc_size(sLen+1);
+            this.buff = chpl_here_alloc(allocSize,
+                                       offset_STR_COPY_DATA): bufferType;
+            c_memcpy(this.buff, s.buff, s.len);
+            this.buff[sLen] = 0;
+            this._size = allocSize;
+          } else {
+            this.buff = s.buff;
+            this._size = s._size;
+          }
+        }
+      }
+    }
+
+
+    // this is implemented only for debugging purposes. Ideally writeThis should
+    // just halt when called on bytes record
+    pragma "no doc"
+    proc writeThis(f) {
```

Comment:
It does not have to happen in this PR, but I would expect the I/O system would support bytes by first updating `_isIoPrimitiveType` to include `bytes` along with `string` and then update the routines handling primitive types (e.g. `_read_text_internal`) to do the right thing with `bytes`, whatever that is.